### PR TITLE
Add basic signal handler and graceful exit

### LINKED ISSUE #1: PID File is supposed to be deleted on heiko stop.
PID File generated in daemon mode is supposed to be deleted via the `context.Release()` call. As of right now, any form of interrupt ( SIGTERM / SIGINT ) doesn't seem to change it's behaviour. 

Also go-daemon seems to be pretty old, could that be the issue?

### PR DESCRIPTION
**Pull request by Samyak2**
Fixes #1 

## Changes

 - Added a signal handler for SIGINT based on the [example](https://github.com/sevlyar/go-daemon/blob/master/examples/cmd/gd-signal-handling/signal-handling.go) in `go-daemon`
    - this ensures that the PID file is deleted
 - Gracefully exit scheduler based on a signal channel (called `stop`)
   - Update: the actual commands are not given a chance for graceful exit, the SSH connections get broken when the daemon exits
   - Update 2: graceful exit is again possible (see 504e7ce), this time with a timeout (default 2s) to ensure that it's not stuck stopping
 - Another signal channel (called `done`) to know when a scheduler exits after
   handling all cleanups
 - In summary: when you do `heiko stop`, heiko signals the schedulers (currently only 1) to exit, waits for them to cleanup and exit, exits the daemon while cleaning up the PID file.

## Bugs

See 8564bd8

~~Need to fix this before merging.~~

 - ~~If a non-exiting task (like `cat`) is running, heiko never~~
   ~~exits as the scheduler keeps waiting for the process to complete xD~~
    - ~~to test it without this bug, either remove that job from the config or comment out lines 19-21 in `internal/daemon/handlers.go`.~~
    - ~~perhaps we need to use context and timeout as described in this [SO answer](https://stackoverflow.com/a/58308642/11199009)~~

### CODE REVIEW

**Comment by MadhavJivrajani** [comment]
Not sure if this matches the use case here, but this could be worth taking a look at, good potential for eliminating a lot of boilerplate -- https://golang.org/pkg/os/signal/#NotifyContext

**Comment by Samyak2** [comment]
Ooh that is interesting. I was trying to use contexts here to make graceful exits, but because the SSH package has no support for contexts and the function is a blocking one, it did provide any benefits of using contexts. The code example there gave me an idea though, thanks!

Although I don't think we can use that particular function because all of the signal handling is done by `go-daemon`, I only provided the handler function.

**Comment by Samyak2** [comment]
I finally figured it out lol. Graceful exit on `heiko stop` works like this now:
 - If there aren't any tasks running, just exit
 - If there are, drain the task queue and wait `Timeout` seconds for them to finish.
 - If the timeout is reached, exit without waiting further

See 504e7ce for details

**Comment by psiayn** [comment]
This PR is probably going to take a while to merge since rn, I am trying to create another repo for easy testing of heiko using docker but don't want to include it in heiko unnecessarily. 

Once I get that up and run; let's try running and building tests using that.

**Comment by Samyak2** [comment]
> This PR is probably going to take a while to merge since rn, I am trying to create another repo for easy testing of heiko using docker but don't want to include it in heiko unnecessarily.
> 
> Once I get that up and run; let's try running and building tests using that.

@psiayn I have a suggestion here. Why not make it in a new module in this repo itself? As long as it's not imported in the main `heiko` module, it won't be built into the binary. This way we can have tests and everything in the same repo and one PR can change both the functionality and the tests, without having to open a new PR elsewhere. What do you think?

---

Also, if there aren't any issues here I can rebase on master and we can merge this PR. You can then develop your tests on a new branch from the latest master with the signal handler included.

**Comment by psiayn** [comment]
That makes sense too ig. 

Go ahead then! Merge the PR and i'll create a new branch for that.

**Review by psiayn — APPROVED** [review]
LGTM!
